amethyst_renderer/src/cam.rs

```diff
@@ -53,25 +53,128 @@ impl From<Projection> for Matrix4<f32> {
 }
 
 /// Camera struct.
-///
-/// TODO: Add more convenience methods, refine API.
 #[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
 pub struct Camera {
+    /// Initial forward vector of the camera.
+    pub base_forward: Vector3<f32>,
+    /// Calculated center point of the camera.
+    center: Vector3<f32>,
+    /// Trigger re-calculations if values were changed.
+    pub dirty: bool,
     /// Location of the camera in three-dimensional space.
     pub eye: Point3<f32>,
     /// Graphical projection of the camera.
     pub proj: Matrix4<f32>,
-    /// Forward vector of the camera.
-    pub forward: Vector3<f32>,
-    /// Right vector of the camera.
-    pub right: Vector3<f32>,
+    /// Rotation of the camera in three-dimensional space.
+    pub rotation: Quaternion<f32>,
     /// Upward elevation vector of the camera.
     pub up: Vector3<f32>,
 }
 
 impl Camera {
+    /// Create a new camera from the given values.
+    pub fn new(
+        projection: Matrix4<f32>,
+        position: Point3<f32>,
+        look_at: Point3<f32>,
+        up: Vector3<f32>
+    ) -> Self {
+        if position == look_at {
+            panic!("Position and LookAt cannot have the same coordinates!");
+        }
+        
+        let base_forward = (look_at - position);
+        Self {
+            base_forward: base_forward,
+            center: (position + base_forward).to_vec(),
+            dirty: false,
+            eye: position,
+            proj: projection,
+            rotation: Quaternion::look_at(base_forward, up).normalize(),
+            up: up,
+        }
+    }
+    
+    /// Move the camera relatively to its current position, but independently from its orientation.
+    pub fn move_absolute(&mut self, direction: Vector3<f32>) -> &mut Self {
```

Comment:
Camera should be just an entity with a pair: `Transform` and `Projection` components.
I like those functions but they all belong to `Transform` component.

Reply:
To know what camera to use there should be special `Resource` with `EntityId`.
In future there will be many viewports and cameras in use. Viewports will have `EntityId` of the camera.

Reply:
I am currently wondering, if those methods shouldn't go into `LocalTransform`? As for multi-cameras, I guess I can prepare that by using an entity with components and then getting the first camera/transform component-pair in the rendering system to shove into the `Scene`. I don't see any viewports, yet, and I think they are out of scope for this PR...

Reply:
For now, I'd just add a CameraEntity resource:

```
pub struct CameraEntity {
    pub entity: Entity,
}
```

and in RenderSystem: `Fetch<'a, CameraEntity>`.

Reply:
And yes, those methods should probably go in LocalTransform.

Reply:
@Rhuagh, `CamearEntity` resource is what I would recomend. And yes, `LocalTransform` is the right place

Reply:
I agree.

Reply:
Why a camera resource? You can just filter for it, and check if it's the only entity with a `Camera` component:

```rust
let mut cams = (&transforms, &cameras).join();
let cam = cams.next();
if cams.next().is_some() {
    // Log error: "Engine only allows one camera at the moment"
}

if let Some(cam) = cam {
    ...
}
```

Reply:
This would at least be more obvious and extensible.

Reply:
I also prefer the `join()`-way. It feels more ECS-y imho. After all, we already have Entities and components. Also, having more than one way to retrieve the correct camera might lead to inconsistencies and bugs later on (which one to use - the resource or the camera?) until the viewports are implemented and define the only way again...

As for the error: I would rather only log a warning. It's not an error we have to handle when there are multiple cameras, the others are just not used. By the way, do we even have a logging system, yet? I think we only have an issue (#124) at the moment... I will leave a `// todo` in place.

Reply:
I'd like to have multiple cameras before multiple viewports. Switching active camera is better than storing data elsewhere and overwrite it.

Reply:
That's why I suggest to have active camera id as a resource.

Reply:
Why not just introduce both at once? It isn't really more work and for now, the solution I suggested is simpler and more obvious.

Reply:
ok, will do. Maybe I can finally push all those changes tonight or tomorrow, so you can kill me :D

Reply:
We like you alive :P

Reply:
@torkleyy so you want to postpone this PR until someone starts working on multiple viewports?

Well, I'd like to keep it simple and I implemented the solution you present here in the scene in order to find the first camera. All others are silently dropped atm (though it might print a warning as soon as the logging system is in place). If @omni-viral needs the resource on top of that, it shouldn't be a problem to implement it. The more I think about it, the more sense it makes to actually use a resource if you want to have multiple cameras now. Later on, we just have to remove the global resource and store the active camera entity on the viewport instead...

Reply:
No. I'd like to do what @omni-viral mentioned in a future PR. I'm not sure if we need multiple cameras now. What would be the use of that? And btw, if we really wanted that the "proper" solution would be to have marker components e.g. `ActiveCamera` (using a `NullStorage`).

Reply:
Yeah, I also wanted to write about an `ActiveCamera` marker, however how would you mark that for multiple VPs later on? Well, I guess, no matter what we do now, it's temporary.

Ah, and btw, maybe he needs them for his own project? idk and idc why he needs it, but if he says it is a requirement for him, why not implement it for now?

Reply:
@torkleyy. I'm not sure when we will implement multiple viewports. But multiple cameras can be implemented right now. So why should we delay. Idc what resource or marker component will point to active camera. Both will be ok. But I think resource is more clear and unambiguous that there is only one resource hence only one active camera.

Reply:
@omni-viral I don't really see any obvious use case for the complication, but if you insist on it, I vote for the component marker.